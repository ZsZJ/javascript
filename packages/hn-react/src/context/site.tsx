import { Site } from 'hn';
import React, { createContext, ReactNode } from 'react';
import globalSite from '../utils/site';

// Create a new context.
const { Provider, Consumer } = createContext(globalSite);

// The site provider is the same as the 'Provider', but changes 'site' to
// 'value'.
export const SiteProvider = ({
  site,
  children,
}: {
  site: Site;
  children: ReactNode;
}) => <Provider value={site} children={children} />;

// The consumer is exported as both Site and SiteConsumer. The SiteConsumer can
// be used if also the Site from 'hn' is imported, so the names don't overlap.
export { Consumer as Site, Consumer as SiteConsumer };

// There is also a HOC withSite that adds the site to the props of the
// component. Source of the HOC for Typescript:
// https://stackoverflow.com/a/50613946/1907875
export interface InjectedSiteProps {
  site: Site;
}

export function withSite<P extends InjectedSiteProps>(
  Component: React.ComponentType<P>,
) {
  return function ComponentWithSite(
    props: Pick<P, Exclude<keyof P, keyof InjectedSiteProps>>,
  ) {
    return (
      <Consumer>{theme => <Component {...props} site={theme} />}</Consumer>
    );
  };
}
