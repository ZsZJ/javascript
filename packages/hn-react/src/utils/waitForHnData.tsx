import * as PropTypes from 'prop-types';
import * as React from 'react';

const asyncBootstrapper = require('react-async-bootstrapper'); // tslint:disable-line:no-var-requires

export default async function(children) {
  const context = {
    state: {
      drupalPage: null,
      entities: [],
    },
  };

  class DrupalPageContextProvider extends React.Component {
    static childContextTypes = {
      hnContext: PropTypes.object,
    };

    getChildContext() {
      return { hnContext: context };
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  const drupalContextProvider = (
    <DrupalPageContextProvider>{children}</DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider, {
    componentWillUnmount: true,
  });

  return drupalContextProvider;
}
