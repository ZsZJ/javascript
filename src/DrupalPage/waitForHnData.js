import React, { Component } from 'react';
import asyncBootstrapper from 'react-async-bootstrapper'
import PropTypes from 'prop-types';

export default async function (children) {

  const context = {};

  class DrupalPageContextProvider extends Component {
    getChildContext() {
      return {hnContext: context};
    }

    render() {
      return React.Children.only(this.props.children);
    }
  }

  DrupalPageContextProvider.childContextTypes = {
    hnContext: PropTypes.object,
  };

  const drupalContextProvider = (
    <DrupalPageContextProvider>
      {children}
    </DrupalPageContextProvider>
  );

  await asyncBootstrapper(drupalContextProvider);

  return drupalContextProvider;

};
