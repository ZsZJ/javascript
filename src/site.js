import { Site } from 'hn';
import getNested from 'get-nested';

const site = new Site({ url: '' });

if(getNested(() => window.__NEXT_DATA__.HN_DATA)) {
  site.data = window.__NEXT_DATA__.HN_DATA;
  site.url = window.__NEXT_DATA__.HN_URL || site.url;
  site.pagesLoading = {
    [window.location.pathname]: Promise.resolve(),
  };
}

export default site;
