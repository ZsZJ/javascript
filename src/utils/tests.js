import React from 'react';

const mapperComponent = jest.fn((props) => {
  return <div className="MyCustomMapperComponent" {...props} />
});

export const mapper = {
  unique_type__unique_bundle: mapperComponent
};

export const uuid = 'unique-uuid';

export const entity = {
  __hn: {
    entity: {
      type: 'unique_type',
      bundle: 'unique_bundle'
    }
  }
};

export function mockSite() {
  class SiteMock {
    getData = jest.fn(() => {
      return entity;
    });
    getPage = jest.fn(async () => {
      return uuid;
    })
  }
  return new SiteMock();
}
