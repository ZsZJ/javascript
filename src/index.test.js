import { Paragraphs } from './index';
import React from 'react';
import renderer from 'react-test-renderer';
import site from './site';
import waitForHnData from './DrupalPage/waitForHnData';
import { mapper, uuid } from './utils/tests';

jest.mock('./site', () => {
  return require('./utils/tests').mockSite();
});
console.warn = jest.fn();

beforeEach(() => {
  console.warn.mockRestore();
  site.getData.mockRestore();
});

describe('Deprecated components', async () => {

  describe('Paragraphs', async () => {

    test('with required props', async() => {
      const component = await waitForHnData(
        <Paragraphs
          mapper={mapper}
          paragraphs={[{target_id: uuid}]}
        />
      );

      expect(renderer.create(component).toJSON()).toMatchSnapshot();

      expect(console.warn).toBeCalledWith('Warning: The prop "paragraphs" is replaced by "entities".');
      expect(console.warn).toBeCalledWith('Warning: The component "Paragraphs" is deprecated, use "EntityListMapper" instead.');
    });

    test('with all props', async() => {
      const component = await waitForHnData(
        <Paragraphs
          mapper={mapper}
          paragraphProps={{
            testProp: 'testPropValue'
          }}
          paragraphs={[{target_id: uuid}]}
          Wrapper={'section'}
          page={{'pageTest': true}}
        />
      );

      expect(renderer.create(component).toJSON()).toMatchSnapshot();

      expect(console.warn).toBeCalledWith('Warning: The prop "paragraphs" is replaced by "entities".');
      expect(console.warn).toBeCalledWith('Warning: The prop "paragraphProps" is replaced by "entityProps".');
      expect(console.warn).toBeCalledWith('Warning: The component "Paragraphs" is deprecated, use "EntityListMapper" instead.');
      expect(console.warn).toBeCalledWith('Warning: The prop "page" is deprecated. Please use "entityProps={{ page }}" instead.');
    });
  });

});
