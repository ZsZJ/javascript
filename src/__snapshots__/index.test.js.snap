// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deprecated components Paragraph with all props 1`] = `
<div
  bundle="unique_type__unique_bundle"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  index={15}
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  testProp="testPropValue"
/>
`;

exports[`Deprecated components Paragraph with required props 1`] = `
<div
  bundle="unique_type__unique_bundle"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  index={0}
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
/>
`;

exports[`Deprecated components Paragraphs with all props 1`] = `
<section>
  <div
    bundle="unique_type__unique_bundle"
    className="MyCustomMapperComponent"
    entity={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle",
            "type": "unique_type",
          },
        },
      }
    }
    index={0}
    page={
      Object {
        "pageTest": true,
      }
    }
    paragraph={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle",
            "type": "unique_type",
          },
        },
      }
    }
    testProp="testPropValue"
  />
</section>
`;

exports[`Deprecated components Paragraphs with required props 1`] = `
<div
  bundle="unique_type__unique_bundle"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
  index={0}
  page={undefined}
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle",
          "type": "unique_type",
        },
      },
    }
  }
/>
`;
